nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/pubsub/ConsumePulsarRecord.java

```diff
@@ -195,240 +197,243 @@ private List<Message<GenericRecord>> getMessages(final Consumer<GenericRecord> c
+        FlowFile flowFile = null;
+        OutputStream rawOut = null;
+        RecordSetWriter writer = null;
+
+        Map<String, String> lastAttributes = null;
+        Message<GenericRecord> lastMessage = null;
+        Map<String, String> currentAttributes = null;
+
+        // Cumulative acks are NOT permitted on Shared subscriptions
+        final boolean shared = isSharedSubscription(context);
+
+        try {
+            for (Message<GenericRecord> msg : groupedMessages) {
+                currentAttributes = getMappedFlowFileAttributes(context, msg);
+                // Introduce an attribute to distinguish between current and previously captured attributes,
+                // particularly when the message originates from a different topic.
+                currentAttributes.put("topicName", msg.getTopicName());
+                // if the current message's mapped attribute values differ from the previous set's,
+                // write out the active record set and clear various references so that we'll start a new one
+                if (lastAttributes != null && !lastAttributes.equals(currentAttributes)) {
+                    WriteResult result = writer.finishRecordSet();
+                    IOUtils.closeQuietly(writer);
+                    IOUtils.closeQuietly(rawOut);
+
+                    if (result != WriteResult.EMPTY) {
+                        flowFile = session.putAllAttributes(flowFile, result.getAttributes());
+                        flowFile = session.putAttribute(flowFile, MSG_COUNT, result.getRecordCount() + "");
+                        session.getProvenanceReporter().receive(flowFile, getPulsarClientService().getPulsarBrokerRootURL() + "/" + consumer.getTopic());
+                        session.transfer(flowFile, REL_SUCCESS);
+                    } else {
+                        session.rollback();
+                    }
+
+                    handleFailures(session, parseFailures, demarcator);
+                    parseFailures.clear();
+
+                    if (!shared) {
+                        acknowledgeCumulative(consumer, lastMessage, async);
+                    }
+
+                    lastAttributes = null;
+                    lastMessage = null;
+                }
+
+                // if there's no record set actively being written, begin one
+                byte[] data = msg.getData();
+                if (lastMessage == null) {
+                    flowFile = session.create();
+                    flowFile = session.putAllAttributes(flowFile, currentAttributes);
+                    if (msg.getReaderSchema().isPresent() && msg.getReaderSchema().get().getSchemaInfo().getType() == SchemaType.AVRO) {
+                        String msgSchema = new String(msg.getReaderSchema().get().getSchemaInfo().getSchema());
+                        flowFile = session.putAttribute(flowFile, "avro.schema", msgSchema);
+                        schema = new SimpleRecordSchema(
```

Comment:
we have to replace the schema effectively here? `schema = getSchema(flowFile, readerFactory, data);` will not do the right thing?